### PR TITLE
boards: olimexino_stm32: Enable CAN support for this board

### PR DESCRIPTION
**Pull request by alexanderwachter**
This PR enables CAN support on the Olimexino STM32 board.

### CODE REVIEW

**Comment by alexanderwachter** [comment]
Splited the commits.

**Review by MaureenHelm — APPROVED** [review]

File: include/dt-bindings/pinctrl/stm32-pinctrlf1.h [hunk 1]
```diff
@@ -85,6 +85,8 @@
 #define STM32_PIN_SPI_SLAVE_MOSI	(STM32_MODE_INPUT | STM32_CNF_IN_FLOAT)
 #define STM32_PIN_SPI_MASTER_MISO	(STM32_MODE_INPUT | STM32_CNF_IN_FLOAT)
 #define STM32_PIN_SPI_SLAVE_MISO	(STM32_MODE_OUTPUT | STM32_CNF_ALT_FUNC | STM32_CNF_PUSH_PULL)
+#define STM32_PIN_CAN_TX		((STM32_MODE_OUTPUT | STM32_CNF_ALT_FUNC | STM32_CNF_PUSH_PULL))
+#define STM32_PIN_CAN_RX		((STM32_MODE_INPUT  | STM32_PUPD_PULL_UP))
```
Comment (erwango) [inline include/dt-bindings/pinctrl/stm32-pinctrlf1.h]: Remove double '(/)'

File: boards/arm/olimexino_stm32/pinmux.c [hunk 2]
```diff
@@ -9,6 +9,7 @@
 #include <init.h>
 #include <drivers/pinmux.h>
 #include <sys/sys_io.h>
+#include <soc.h>
```
Comment (erwango) [inline boards/arm/olimexino_stm32/pinmux.c]: Is that actually required?

**Review by erwango — CHANGES_REQUESTED** [review]
Please fix double sign-off in second commit header.
Couple of minor comments otherwise

**Comment by walzert** [comment]
Tested the samples/drivers/CAN with the device an the output is the same as the sample output.

File: boards/arm/olimexino_stm32/pinmux.c (hunk 2 above)
Comment (alexanderwachter) [reply boards/arm/olimexino_stm32/pinmux.c]: Yes, for the pinmuxing register

File: boards/arm/olimexino_stm32/pinmux.c [hunk 3]
```diff
@@ -57,13 +58,20 @@ static const struct pin_config pinconf[] = {
...
+#endif /* CONFIG_CAN_1 */
 };
 
 static int pinmux_stm32_init(struct device *port)
 {
 	ARG_UNUSED(port);
 
 	stm32_setup_pins(pinconf, ARRAY_SIZE(pinconf));
+#ifdef CONFIG_CAN_1
+	AFIO->MAPR |= AFIO_MAPR_CAN_REMAP_REMAP2;
```
Comment (erwango) [inline boards/arm/olimexino_stm32/pinmux.c]: Had overlook this line.
Can you add a comment stating this is due to this particular pin cofniguration ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/olimexino_stm32/pinmux.c (hunk 3 above)
Comment (alexanderwachter) [reply boards/arm/olimexino_stm32/pinmux.c]: Done

File: include/dt-bindings/pinctrl/stm32-pinctrlf1.h (hunk 1 above)
Comment (alexanderwachter) [reply include/dt-bindings/pinctrl/stm32-pinctrlf1.h]: Done

File: include/dt-bindings/pinctrl/stm32-pinctrlf1.h [hunk 4]
```diff
@@ -85,8 +85,8 @@
 #define STM32_PIN_SPI_SLAVE_MOSI	(STM32_MODE_INPUT | STM32_CNF_IN_FLOAT)
 #define STM32_PIN_SPI_MASTER_MISO	(STM32_MODE_INPUT | STM32_CNF_IN_FLOAT)
 #define STM32_PIN_SPI_SLAVE_MISO	(STM32_MODE_OUTPUT | STM32_CNF_ALT_FUNC | STM32_CNF_PUSH_PULL)
-#define STM32_PIN_CAN_TX		((STM32_MODE_OUTPUT | STM32_CNF_ALT_FUNC | STM32_CNF_PUSH_PULL))
-#define STM32_PIN_CAN_RX		((STM32_MODE_INPUT  | STM32_PUPD_PULL_UP))
+#define STM32_PIN_CAN_TX		(STM32_MODE_OUTPUT | STM32_CNF_ALT_FUNC | STM32_CNF_PUSH_PULL)
+#define STM32_PIN_CAN_RX		(STM32_MODE_INPUT  | STM32_PUPD_PULL_UP)
```
Comment (erwango) [inline include/dt-bindings/pinctrl/stm32-pinctrlf1.h]: Can you push this previous commit ?

**Review by erwango — CHANGES_REQUESTED** [review]
Double sign-off still here.
And minor comment

**Comment by ydamigos** [comment]
@alexanderwachter Does CAN need special setup on olimexino_stm32 board to work? I tried `samples/drivers/CAN/` and `tests/drivers/can/api/` but none of them worked.

Serial output of `tests/drivers/can/api/`:
```
E: Failed to exit sleep mode                                         |
***** Booting Zephyr OS build zephyr-v2.0.0-1685-g4499b5844336 *****  

    Assertion failed at ZEPHYR_BASE/tests/drivers/can/api/src/main.c:602: test_main: (can_dev is NULL)
Device not found
ASSERTION FAIL [!(thread->base.user_options & ((1UL << (0))))] @ ZEPHYR_BASE/arch/arm/core/cortex_m/thread_abort.c7
        essential thread aborted
E: r0/a1:  0x00000004  r1/a2:  0x00000025  r2/a3:  0xffffffbf
E: r3/a4:  0x0800829e r12/ip:  0x00000009 r14/lr:  0x080031d1
E:  xpsr:  0x61000000
E: Faulting instruction address (r15/pc): 0x0800614e
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic
E: Current thread: 0x20000334 (main)
E: Halting system
```
Serial output of `samples/drivers/CAN/`:
```
***** Booting Zephyr OS build zephyr-v2.0.0-1685-g4499b5844336 ***** |
CAN: Device driver not found.
```

File: boards/arm/olimexino_stm32/doc/index.rst [hunk 5]
```diff
@@ -334,6 +348,15 @@ through its mini USB connector.
 - USB_DM : PA11
 - USB_DP : PA12
 
+CAN
+===
+
+OLIMEXINO-STM32 board has a CAN interface with transceiver on board.
+CAN is accessible trough a screw terminal.
```
Comment (dbkinder) [inline boards/arm/olimexino_stm32/doc/index.rst]: change ``trough`` to ``through``

**Review by dbkinder — CHANGES_REQUESTED** [review]
one misspelling, otherwise docs LGTM.

**Comment by alexanderwachter** [comment]
> @alexanderwachter Does CAN need special setup on olimexino_stm32 board to work? I tried `samples/drivers/CAN/` and `tests/drivers/can/api/` but none of them worked.

Actually not. It seems that the Controller is not initialized. This has nothing to do with the board but with the SoC.
Could you turn on logging and send me the log?

File: boards/arm/olimexino_stm32/doc/index.rst (hunk 5 above)
Comment (alexanderwachter) [reply boards/arm/olimexino_stm32/doc/index.rst]: Thanks for reviewing!
Changed.

File: include/dt-bindings/pinctrl/stm32-pinctrlf1.h (hunk 4 above)
Comment (alexanderwachter) [reply include/dt-bindings/pinctrl/stm32-pinctrlf1.h]: Done.

**Review by dbkinder — APPROVED** [review]
+1 for docs, thanks.

**Comment by walzert** [comment]
west build -b olimexino_stm32 tests/drivers/can/api --pristine
../../stm32flash/stm32flash -w build/zephyr/zephyr.bin -v -g 0x8 /dev/ttyUSB0 

has following result:


�I: Init of CAN_1 done
***** Booting Zephyr OS build zephyr-v2.0.0-1685-g8cb622e369eb *****
Running test suite can_driver
===================================================================
starting test - test_set_loopback
PASS - test_set_loopback
===================================================================
starting test - test_send_and_forget
PASS - test_send_and_forget
===================================================================
starting test - test_filter_attach
PASS - test_filter_attach
===================================================================
starting test - test_receive_timeout
PASS - test_receive_timeout
===================================================================
starting test - test_send_callback
PASS - test_send_callback
===================================================================
starting test - test_send_receive_std
PASS - test_send_receive_std
===================================================================
starting test - test_send_receive_ext
PASS - test_send_receive_ext
===================================================================
starting test - test_send_receive_std_masked
PASS - test_send_receive_std_masked
===================================================================
starting test - test_send_receive_ext_masked
PASS - test_send_receive_ext_masked
===================================================================
starting test - test_send_receive_buffer
PASS - test_send_receive_buffer
===================================================================
starting test - test_send_receive_wrong_id
PASS - test_send_receive_wrong_id
===================================================================
Test suite can_driver succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL

**Review by erwango — APPROVED** [review]
Approved from my side (but would be nice to fix the double sign-off in second commit.)
I let @ydamigos commenting on the testing side.

**Comment by ydamigos** [comment]
@walzert So I am doing something wrong. How do you setup your board? For example, do you connect CAN_RX to CAN_TX?  Do I need another device with CAN support?

**Comment by walzert** [comment]
@ydamigos there is the loopback_mode. With that the example is working without a other can device. I just build the bin and flashed it. With a second device I can test it on monday.

**Comment by ydamigos** [comment]
@walzert Thanks for the tip. Enabling loopback mode `tests/drivers/can/api/` is working.

**Review by ydamigos — APPROVED** [review]

**Comment by galak** [comment]
approved by TSC for merge in 2.1

**Pull request merged**